File: rust_image/android/src/main/kotlin/com/flutter_rust_bridge/rust_image/RustImageMediaPipeAnalyzer.kt
package com.flutter_rust_bridge.rust_image

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.facelandmarker.FaceLandmarker
import com.google.mediapipe.tasks.vision.facelandmarker.FaceLandmarkerResult
import com.google.mediapipe.tasks.vision.imagesegmenter.ImageSegmenter
import com.google.mediapipe.tasks.vision.imagesegmenter.ImageSegmenterResult
import java.io.File
import java.nio.ByteBuffer

/**
 * MediaPipe 468-point face mesh when models are downloaded (Nexus A).
 */
internal object RustImageMediaPipeAnalyzer {
    private const val MIN_LANDMARKS = 468
    private const val FACE_OVAL_COUNT = 36

    fun modelsReady(modelDir: String?): Boolean {
        if (modelDir.isNullOrBlank()) return false
        val base = File(modelDir)
        return File(base, "face_landmarker.task").exists() &&
            File(base, "selfie_segmenter.tflite").exists()
    }

    fun analyze(
        imageBytes: ByteArray,
        pixelFormat: String,
        targetWidth: Int,
        targetHeight: Int,
        modelDir: String,
    ): Map<String, Any> {
        val bitmap = decodeBitmap(imageBytes, pixelFormat, targetWidth, targetHeight)
            ?: throw IllegalArgumentException("Could not decode image")
        val scaled = Bitmap.createScaledBitmap(bitmap, targetWidth, targetHeight, true)

        val facePath = File(modelDir, "face_landmarker.task").absolutePath
        val segPath = File(modelDir, "selfie_segmenter.tflite").absolutePath

        val landmarks = detectLandmarks(scaled, facePath, targetWidth, targetHeight)
        val mask = segmentSelfie(scaled, segPath, targetWidth, targetHeight)

        val confidence = if (landmarks.size >= MIN_LANDMARKS) 0.98f else 0f
        val landmarkMaps = landmarks.map {
            mapOf("x" to it.first.toDouble(), "y" to it.second.toDouble(), "z" to 0.0)
        }

        return mapOf(
            "landmarks" to landmarkMaps,
            "confidence" to confidence.toDouble(),
            "faceContourCount" to FACE_OVAL_COUNT,
            "regionCounts" to emptyList<Int>(),
            "mask" to mapOf(
                "width" to targetWidth,
                "height" to targetHeight,
                "bytes" to mask,
            ),
            "meshKind" to "mediapipe468",
        )
    }

    private fun decodeBitmap(
        bytes: ByteArray,
        pixelFormat: String,
        width: Int,
        height: Int,
    ): Bitmap? {
        if (pixelFormat == "rgba" && width > 0 && height > 0 && bytes.size >= width * height * 4) {
            return Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888).also { bmp ->
                bmp.copyPixelsFromBuffer(ByteBuffer.wrap(bytes))
            }
        }
        return BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
    }

    private fun detectLandmarks(
        bitmap: Bitmap,
        modelPath: String,
        width: Int,
        height: Int,
    ): List<Pair<Float, Float>> {
        val baseOptions = BaseOptions.builder().setModelAssetPath(modelPath).build()
        val options = FaceLandmarker.FaceLandmarkerOptions.builder()
            .setBaseOptions(baseOptions)
            .setRunningMode(RunningMode.IMAGE)
            .setNumFaces(1)
            .build()
        val landmarker = FaceLandmarker.createFromOptions(RustImageAppContext.get(), options)
        val mpImage = BitmapImageBuilder(bitmap).build()
        val result: FaceLandmarkerResult = landmarker.detect(mpImage)
        landmarker.close()

        val face = result.faceLandmarks().firstOrNull() ?: return emptyList()
        return face.map { lm ->
            Pair(lm.x(), lm.y())
        }
    }

    private fun segmentSelfie(
        bitmap: Bitmap,
        modelPath: String,
        width: Int,
        height: Int,
    ): ByteArray {
        val baseOptions = BaseOptions.builder().setModelAssetPath(modelPath).build()
        val options = ImageSegmenter.ImageSegmenterOptions.builder()
            .setBaseOptions(baseOptions)
            .setRunningMode(RunningMode.IMAGE)
            .setOutputCategoryMask(true)
            .build()
        val segmenter = ImageSegmenter.createFromOptions(RustImageAppContext.get(), options)
        val mpImage = BitmapImageBuilder(bitmap).build()
        val result: ImageSegmenterResult = segmenter.segment(mpImage)
        segmenter.close()

        val mask = result.categoryMask().get()
        val maskWidth = mask.width
        val maskHeight = mask.height
        val buffer = mask.buffer
        val out = ByteArray(width * height)

        for (y in 0 until height) {
            val sy = y * maskHeight / height
            for (x in 0 until width) {
                val sx = x * maskWidth / width
                val idx = sy * maskWidth + sx
                val confidence = buffer.get(idx).toInt() and 0xFF
                out[y * width + x] = if (confidence > 0) 255.toByte() else 0
            }
        }
        return out
    }
}

File: packages/rust_image_core/ios/Classes/Darwin/RustImageMediaPipeAnalyzer.swift
import CoreGraphics
import Foundation

#if canImport(MediaPipeTasksVision)
import MediaPipeTasksVision
#endif

#if canImport(UIKit)
import UIKit
#endif
#if canImport(UIKit)
import UIKit
#endif

/// MediaPipe Face Landmarker (468 pts) + selfie segmenter TFLite when models are on disk.
enum RustImageMediaPipeAnalyzer {
  private static let minLandmarksForValid = 468

  static func modelsReady(at modelDir: String?) -> Bool {
    guard let dir = modelDir, !dir.isEmpty else { return false }
    let base = URL(fileURLWithPath: dir, isDirectory: true)
    let face = base.appendingPathComponent("face_landmarker.task")
    let seg = base.appendingPathComponent("selfie_segmenter.tflite")
    return FileManager.default.fileExists(atPath: face.path)
      && FileManager.default.fileExists(atPath: seg.path)
  }

  static func analyze(
    imageData: Data,
    pixelFormat: String,
    targetWidth: Int,
    targetHeight: Int,
    modelDir: String
  ) throws -> [String: Any] {
    #if canImport(MediaPipeTasksVision)
    guard modelsReady(at: modelDir) else {
      throw NSError(
        domain: "rust_image",
        code: 10,
        userInfo: [NSLocalizedDescriptionKey: "MediaPipe models not found"]
      )
    }
    guard let cgImage = RustImageImageDecode.cgImage(
      from: imageData,
      pixelFormat: pixelFormat,
      width: targetWidth,
      height: targetHeight
    ) else {
      throw NSError(domain: "rust_image", code: 1, userInfo: [NSLocalizedDescriptionKey: "Could not decode image"])
    }
    let tw = max(1, targetWidth)
    let th = max(1, targetHeight)
    guard let scaled = resize(cgImage, width: tw, height: th) else {
      throw NSError(domain: "rust_image", code: 2, userInfo: [NSLocalizedDescriptionKey: "Could not scale image"])
    }

    let facePath = URL(fileURLWithPath: modelDir).appendingPathComponent("face_landmarker.task").path
    let segPath = URL(fileURLWithPath: modelDir).appendingPathComponent("selfie_segmenter.tflite").path

    let landmarks = try detectLandmarks(cgImage: scaled, modelPath: facePath, width: tw, height: th)
    let mask = try segmentSelfie(cgImage: scaled, modelPath: segPath, width: tw, height: th)

    let confidence: Float = landmarks.count >= minLandmarksForValid ? 0.98 : 0.0
    let landmarkMaps: [[String: Double]] = landmarks.map {
      ["x": Double($0.x), "y": Double($0.y), "z": Double($0.z)]
    }

    return [
      "landmarks": landmarkMaps,
      "confidence": confidence,
      "faceContourCount": FACE_OVAL_COUNT,
      "regionCounts": [Int](),
      "mask": [
        "width": tw,
        "height": th,
        "bytes": FlutterStandardTypedData(bytes: mask),
      ],
      "meshKind": "mediapipe468",
    ]
    #else
    throw NSError(
      domain: "rust_image",
      code: 11,
      userInfo: [NSLocalizedDescriptionKey: "MediaPipeTasksVision not linked"]
    )
    #endif
  }

  private static let FACE_OVAL_COUNT = 36

  #if canImport(MediaPipeTasksVision)
  private static func makeMPImage(cgImage: CGImage) throws -> MPImage {
    #if canImport(UIKit)
    return try MPImage(uiImage: UIImage(cgImage: cgImage))
    #else
    throw NSError(domain: "rust_image", code: 3, userInfo: [NSLocalizedDescriptionKey: "Unsupported platform for MPImage"])
    #endif
  }

  private static func detectLandmarks(
    cgImage: CGImage,
    modelPath: String,
    width: Int,
    height: Int
  ) throws -> [(x: Float, y: Float, z: Float)] {
    let landmarker = try FaceLandmarker(modelPath: modelPath)
    let mpImage = try makeMPImage(cgImage: cgImage)
    let result = try landmarker.detect(image: mpImage)
    guard let face = result.faceLandmarks.first else { return [] }

    return face.map { lm in
      (
        x: Float(lm.x),
        y: Float(lm.y),
        z: Float(lm.z)
      )
    }
  }

  private static func segmentSelfie(
    cgImage: CGImage,
    modelPath: String,
    width: Int,
    height: Int
  ) throws -> Data {
    let segmenterOptions = ImageSegmenterOptions()
    segmenterOptions.baseOptions = BaseOptions()
    segmenterOptions.baseOptions.modelAssetPath = modelPath
    segmenterOptions.runningMode = .image
    segmenterOptions.shouldOutputCategoryMask = true
    segmenterOptions.shouldOutputConfidenceMasks = false
    let segmenter = try ImageSegmenter(options: segmenterOptions)
    let mpImage = try makeMPImage(cgImage: cgImage)
    let result = try segmenter.segment(image: mpImage)
    guard let mask = result.categoryMask else {
      return Data(repeating: 0, count: width * height)
    }
    let w = mask.width
    let h = mask.height
    var out = Data(count: width * height)
    out.withUnsafeMutableBytes { dstRaw in
      guard let dst = dstRaw.baseAddress?.assumingMemoryBound(to: UInt8.self) else { return }
      for y in 0..<height {
        let sy = y * h / height
        for x in 0..<width {
          let sx = x * w / width
          let idx = sy * w + sx
          let v = mask.uint8Data[idx]
          dst[y * width + x] = v > 0 ? 255 : 0
        }
      }
    }
    return out
  }
  #endif

  private static func resize(_ image: CGImage, width: Int, height: Int) -> CGImage? {
    let colorSpace = CGColorSpaceCreateDeviceRGB()
    guard let ctx = CGContext(
      data: nil,
      width: width,
      height: height,
      bitsPerComponent: 8,
      bytesPerRow: width * 4,
      space: colorSpace,
      bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
    ) else { return nil }
    ctx.interpolationQuality = .high
    ctx.draw(image, in: CGRect(x: 0, y: 0, width: width, height: height))
    return ctx.makeImage()
  }
}

#if canImport(FlutterMacOS)
import FlutterMacOS
#else
import Flutter
#endif
